sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/IOITHelper.java

```diff
@@ -36,4 +40,54 @@ public static String getHashForRecordCount(int recordCount, Map<Integer, String>
     }
     return hash;
   }
+
+  /**
+   * Interface for passing method to executeWithRetry method.
+   */
+  public interface RetryFunction {
+    void run() throws Exception;
+  }
+
+  public static void executeWithRetry(RetryFunction function)
+      throws Exception {
+    int maxAttempts = 3;
+    long minDelay = 10_000;
+    executeWithRetry(maxAttempts, minDelay, function);
+  }
+
+  public static void executeWithRetry(int maxAttempts, long minDelay, RetryFunction function)
+      throws Exception {
+    ArrayList<Exception> errorList = new ArrayList<Exception>();
+    int attempts = 0;
+    long delay = minDelay;
+
+    while (attempts < maxAttempts) {
+      try {
+        function.run();
+        return;
+      } catch (Exception e) {
+        LOG.error(
+            "Attempt #{} of {} threw an exception: {} after {} ms",
+            attempts + 1,
+            maxAttempts,
+            e.getMessage(),
+            delay);
+        errorList.add(e);
+
+        if (attempts == maxAttempts - 1) {
```

Comment:
Well, it depends how we're counting. right now we are starting from attempts = 0 and in while there is ` while (attempts < maxAttempts)`.
I can change the start value of attempts to 1, then as you said change this if to `if (attempts == maxAttempts)`.  I guess it can look better.